design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+### Admin operations
+
+#### `SHOW TOPICS`
+
+```
+  CompletableFuture<List<TopicInfo>> listTopics();
+```
+with
+```
+public interface TopicInfo {
+
+  String getName();
+
+  int getPartitions();
+  
+  List<Integer> getReplicasPerPartition();
+
+}
+```
+
+#### `SHOW <STREAMS/TABLES>`
+
+```
+  CompletableFuture<List<StreamInfo>> listStreams();
+
+  CompletableFuture<List<TableInfo>> listTables();
+```
+with
+```
+public interface StreamInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+  
+}
+```
+and
+```
+public interface TableInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+
+  boolean isWindowed();
+  
+}
+```
+
+I'm not sure whether it makes more sense for `StreamInfo#getFormat()` and `TableInfo#getFormat()` to return a string or an enum value.
+The latter would make it easier for the user to know the possible values, but we'd have to keep the list up to date and would also sacrifice forward compatibility.
+
+#### `SHOW QUERIES`
+
+```
+  CompletableFuture<List<QueryInfo>> listQueries();
+```
+with
+```
+public interface QueryInfo {
+
+  boolean isPersistentQuery();
+
+  boolean isPushQuery();
+
+  /**
+   * Query ID, used for control operations such as terminating the query
+   */
+  String getId();
+
+  String getSql();
+
+  /**
+   * Name of sink, for a persistent query. Else, null.
+   */
+  String getSink();
+
+  /**
+   * Name of sink topic, for a persistent query. Else, null.
+   */
+  String getSinkTopic();
+  
+  /**
+   * Map of query state ("RUNNING", "ERROR", or "UNRESPONSIVE") to the number of ksqlDB servers
+   * on which this query is in each state.
+   */
+  Map<String, Integer> getServerStatusCounts();
+
+}
+```
+
+Again, not sure whether it makes more sense to represent query states as strings, or to introduce an Enum and have `getServerStatusCounts()` return an EnumMap instead.
+
+Does having `getSink()` and `getSinkTopic()` return null in the case of push queries make sense, or would we prefer an empty string or an Optional instead?
+
+#### `SHOW CONNECTORS`
+
+```
+  CompletableFuture<ConnectorList> listConnectors();
+```
+with
+```
+public interface ConnectorList {
+  
+  List<ConnectorInfo> getConnectors();
+  
+  /**
+   * Any warnings returned by the server as a result of listing connectors.
+   */
+  List<String> getWarnings();
+
+}
+```
+and
+```
+public interface ConnectorInfo {
+
+  enum ConnectorType {
+    SOURCE,
+    SINK,
+    UNKNOWN;
+  }
+
+  String getName();
+
+  ConnectorType getType();
+
+  String getClassName();
+
+  String getState();
+
+}
+```
+
+I don't love that the introduction of `ConnectorList` wrapped around `List<ConnectorInfo>` breaks the pattern established by `SHOW TOPICS`/`SHOW <STREAMS/TABLES>`/`SHOW QUERIES` but it seems important to propagate any server warnings to the user so the trade-off is worth it IMO.
+
+### Terminate push query
+
+```
+  CompletableFuture<Void> terminatePushQuery(String queryId);
```

Comment:
why is the return of the CompleteableFuture void? seems like we'd want to know if our terminate succeeded

Reply:
If the CompletableFuture succeeds, the request succeeded. If the CompletableFuture completes exceptionally, the request failed.

Reply:
in my personal preference, it's better to have it return an object that can contain more useful information (e.g. structured error messages) and not rely on exceptions

Reply:
The CompletableFuture way of doing things is to use an exception for failure. If you call .get() on the CF, the exception will be thrown. Of course, if you want extra information you can create your own exception types with all the information that you require.

Reply:
if that's a standard i'm happy to defer to that - i havne't used the new CompletableFuture much tbh